File: sdk/python/src/pyro_sdk/sandbox.py
"""Sandbox class — represents a running Pyro sandbox."""

from __future__ import annotations

from typing import TYPE_CHECKING, AsyncIterator

import httpx

from pyro_sdk.errors import ExecError, PyroError, SandboxNotFoundError
from pyro_sdk.models import ExecResult, SandboxInfo

if TYPE_CHECKING:
    from pyro_sdk.client import Pyro


class Sandbox:
    """A running Pyro sandbox. Use Pyro.sandbox.create() to get one."""

    def __init__(self, client: Pyro, info: SandboxInfo):
        self._client = client
        self.info = info
        self.id = info.id

    async def exec(
        self,
        command: list[str],
        *,
        env: dict[str, str] | None = None,
        workdir: str | None = None,
        timeout: int | None = None,
    ) -> ExecResult:
        """Execute a command in the sandbox."""
        body: dict = {"command": command}
        if env:
            body["env"] = env
        if workdir:
            body["workdir"] = workdir
        if timeout:
            body["timeout"] = timeout

        data = await self._client._request(
            "POST", f"/sandboxes/{self.id}/exec", json=body
        )
        return ExecResult(
            exit_code=data["exit_code"],
            stdout=data.get("stdout", ""),
            stderr=data.get("stderr", ""),
        )

    async def run(self, code: str, *, language: str | None = None) -> ExecResult:
        """Run code in the sandbox. Infers interpreter from image if not specified."""
        lang = language or self._infer_language()
        if lang == "python":
            return await self.exec(["python3", "-c", code])
        elif lang == "node":
            return await self.exec(["node", "-e", code])
        else:
            return await self.exec(["sh", "-c", code])

    async def write_file(self, path: str, content: str | bytes) -> int:
        """Write a file into the sandbox. Returns bytes written."""
        headers = {}
        if isinstance(content, bytes):
            headers["Content-Type"] = "application/octet-stream"
        else:
            headers["Content-Type"] = "text/plain"
            content = content.encode()

        resp = await self._client._http.put(
            f"{self._client._base_url}/api/sandboxes/{self.id}/files{path}",
            content=content,
            headers={**self._client._headers(), **headers},
        )
        self._client._check_response(resp)
        data = resp.json()
        return data.get("bytes_written", len(content))

    async def read_file(self, path: str) -> bytes:
        """Read a file from the sandbox."""
        resp = await self._client._http.get(
            f"{self._client._base_url}/api/sandboxes/{self.id}/files{path}",
            headers=self._client._headers(),
        )
        self._client._check_response(resp)
        return resp.content

    async def stop(self) -> None:
        """Destroy this sandbox."""
        await self._client._request("DELETE", f"/sandboxes/{self.id}")

    async def status(self) -> SandboxInfo:
        """Get current sandbox status."""
        data = await self._client._request("GET", f"/sandboxes/{self.id}")
        self.info = SandboxInfo.from_dict(data)
        return self.info

    def _infer_language(self) -> str:
        image = self.info.image.lower()
        if "python" in image:
            return "python"
        if "node" in image:
            return "node"
        return "shell"

    async def __aenter__(self):
        return self

    async def __aexit__(self, *exc):
        await self.stop()

    def __repr__(self) -> str:
        return f"Sandbox(id={self.id!r}, image={self.info.image!r}, state={self.info.state!r})"

File: sdk/typescript/src/sandbox.ts
import type { Pyro } from "./client.js";
import type { ExecOptions, ExecResult, SandboxInfo } from "./types.js";

/** A running Pyro sandbox. Use pyro.sandbox.create() to get one. */
export class Sandbox {
  readonly id: string;
  info: SandboxInfo;
  private client: Pyro;

  constructor(client: Pyro, info: SandboxInfo) {
    this.client = client;
    this.info = info;
    this.id = info.id;
  }

  /** Execute a command in the sandbox. */
  async exec(command: string[], options?: ExecOptions): Promise<ExecResult> {
    const body: Record<string, unknown> = { command };
    if (options?.env) body.env = options.env;
    if (options?.workdir) body.workdir = options.workdir;
    if (options?.timeout) body.timeout = options.timeout;

    const data = await this.client.request<Record<string, unknown>>(
      "POST",
      `/sandboxes/${this.id}/exec`,
      body,
    );
    return {
      exitCode: data.exit_code as number,
      stdout: (data.stdout as string) ?? "",
      stderr: (data.stderr as string) ?? "",
    };
  }

  /** Run code in the sandbox. Infers interpreter from image. */
  async run(code: string, language?: string): Promise<ExecResult> {
    const lang = language ?? this.inferLanguage();
    if (lang === "python") {
      return this.exec(["python3", "-c", code]);
    } else if (lang === "node") {
      return this.exec(["node", "-e", code]);
    }
    return this.exec(["sh", "-c", code]);
  }

  /** Write a file into the sandbox. */
  async writeFile(path: string, content: string | Uint8Array): Promise<void> {
    const isText = typeof content === "string";
    const resp = await fetch(
      `${this.client.baseUrl}/api/sandboxes/${this.id}/files${path}`,
      {
        method: "PUT",
        headers: {
          ...this.client.headers(),
          "Content-Type": isText ? "text/plain" : "application/octet-stream",
        },
        body: isText ? (content as string) : Buffer.from(content as Uint8Array),
      },
    );
    this.client.checkResponse(resp);
  }

  /** Read a file from the sandbox. */
  async readFile(path: string): Promise<Uint8Array> {
    const resp = await fetch(
      `${this.client.baseUrl}/api/sandboxes/${this.id}/files${path}`,
      { headers: this.client.headers() },
    );
    this.client.checkResponse(resp);
    return new Uint8Array(await resp.arrayBuffer());
  }

  /** Destroy this sandbox. */
  async stop(): Promise<void> {
    await this.client.request("DELETE", `/sandboxes/${this.id}`);
  }

  /** Get current sandbox status. */
  async status(): Promise<SandboxInfo> {
    const data = await this.client.request<Record<string, unknown>>(
      "GET",
      `/sandboxes/${this.id}`,
    );
    this.info = parseSandboxInfo(data);
    return this.info;
  }

  private inferLanguage(): string {
    const image = this.info.image.toLowerCase();
    if (image.includes("python")) return "python";
    if (image.includes("node")) return "node";
    return "shell";
  }
}

export function parseSandboxInfo(data: Record<string, unknown>): SandboxInfo {
  return {
    id: data.id as string,
    image: data.image as string,
    state: data.state as string,
    pid: data.pid as number,
    vsockCid: data.vsock_cid as number,
    createdAt: data.created_at as string,
    expiresAt: data.expires_at as string,
    apiKeyId: data.api_key_id as string | undefined,
    vcpu: data.vcpu as number | undefined,
    memMib: data.mem_mib as number | undefined,
  };
}
